sos/report/plugins/ceph_osd.py

```diff
@@ -6,37 +6,74 @@
 #
 # See the LICENSE file in the source distribution for further information.
 
+import os
+
 from sos.report.plugins import Plugin, RedHatPlugin, UbuntuPlugin
 
 
 class CephOSD(Plugin, RedHatPlugin, UbuntuPlugin):
+    """
+    This plugin is for capturing information from Ceph OSD nodes. While the
+    majority of this plugin should be version agnotics, several collections are
+    dependent upon the version of Ceph installed. Versions that correlate to
+    RHCS 4 or RHCS 5 are explicitly handled for differences such as those
+    pertaining to log locations on the host filesystem.
+
+    Note that while this plugin will activate based on the presence of Ceph
+    containers, commands are run directly on the host as those containers are
+    often not configured to successfully run the `ceph` commands collected by
+    this plugin. These commands are majorily `ceph daemon` commands that will
+    reference discovered admin sockets under /var/run/ceph.
+    """
 
     short_desc = 'CEPH osd'
 
     plugin_name = 'ceph_osd'
     profiles = ('storage', 'virt', 'container')
-    containers = ('ceph-(.*)?osd.*',)
-    files = ('/var/lib/ceph/osd/',)
+    containers = ('ceph-(.*-)?osd.*',)
+    files = ('/var/lib/ceph/osd/', '/var/lib/ceph/*/osd*')
 
     def setup(self):
+
+        self.ceph_version = self.get_ceph_version()
+
+        logdir = '/var/log/ceph'
+        libdir = '/var/lib/ceph'
+        rundir = '/run/ceph'
+
+        if self.ceph_version >= 16:
+            logdir += '/*'
+            libdir += '/*'
+            rundir += '/*'
+
         self.add_file_tags({
-            '/var/log/ceph/ceph-osd.*.log': 'ceph_osd_log',
+            f"{logdir}/ceph-(.*-)?osd.*.log": 'ceph_osd_log',
         })
 
+        self.add_forbidden_path([
+            "/etc/ceph/*keyring*",
+            f"{libdir}/*keyring*",
+            f"{libdir}/**/*keyring*",
+            # Excludes temporary ceph-osd mount location like
+            # /var/lib/ceph/tmp/mnt.XXXX from sos collection.
+            f"{libdir}/tmp/*mnt*",
+            "/etc/ceph/*bindpass*"
+        ])
+
         # Only collect OSD specific files
         self.add_copy_spec([
-            "/run/ceph/ceph-osd*",
-            "/var/lib/ceph/osd/*/kv_backend",
-            "/var/log/ceph/ceph-osd*.log",
-            "/var/log/ceph/ceph-volume*.log",
+            f"{rundir}/ceph-osd*",
+            f"{libdir}/**/kv_backend",
```

Comment:
The `**` is intentional as the `kv_backend` can be in any subdir, right?

Reply:
Correct.